keps/sig-api-machinery/3488-cel-admission-control/README.md

```diff
@@ -0,0 +1,2636 @@
+- https://github.com/open-policy-agent/gatekeeper/blob/a1add93b0beb5c48eb92a6a2eb5ee7d21551a1b6/pkg/mutation/match/match.go#L22-L56
+
+Alternative considered: Define match criteria with CEL expressions
+
+Pros:
+
+- CEL expressions can be used to declare far more sophisticated match criteria.
+
+Cons:
+
+- CEL expressions can evalaute to an error while API types only evaluate to
+  "matches" / "does not match" result.
+- We intend to allow policy configuration resources declare match criteria,
+  making it difficult to pre-compile or type-check CEL expressions.
+- Makes it more difficult to offer support for a expression language other than
+  CEL in the future.
+
+###### Exclude matching
+
+Benefits:
+
+- Exclude rules make it easier for policy definition authors to do things like
+  exclude kube-system or other special purpose namespaces from policies that apply
+  to all "normal" namespaces.
+- Default rules make it easier to apply principal of least privilege to
+  policies. I.e. ensure that newly created resources default to a least
+  privilege state until configured otherwise.
+
+Exclude rules can be modeled after the approach exstablished in Audit where
+level=None on a match rule can be used to exlude any rules beneath it. E.g.
+`effect: Exclude`.
+
+If we want this, it is essential that we introduce it early, the exclude inverts
+the semantics of a match, adding it in a later release is _very_ incompatible.
+
+#### Reporting violations to Clients
+
+This section focuses on how information is reported back to clients in
+when validations fail.
+
+<<[UNRESOLVED ]>>
+Lots of open questions about this part of the KEP that need resolution:
+
+- Should enforcement be configured on the policy definition, the policy
+  configuration, or both (with some sort of override mechanism)?
+- Are Deny, Warn and Audit the right set of enforcements?
+- Should there also be an Ignore enforcement level (disablement option)?
+- Recording audit annotations requires an audit key. How to handle? E.g. if a
+  policy defines multiple validations and the configuration is set to
+  enforcement=Audit, what audit keys does the policy use?
+- If policy definitions can set enforcement, should it be per-validation
+  expression or per-policy?
+- If policy configurations can configure enforcement, is there only a single
+  field with a single enforcement option that can be set, or are there more fine
+  grained controls for handling each of the validations that can fail in a
+  policy?
+<<[/UNRESOLVED]>>
+
+Goals:
+
+- Feature parity with `AdmissionReview`
+  - Support allow/deny result, warnings, audit annotations
+  - Support for reasons/codes
+- Ability to format message strings
+
+High level proposal:
+
+- Each policy may set a `denyReason` and/or `denyCode`
+- Each validation may define a message:
+  - `message` for plain strings
+  - `messageExpression` for a CEL expression evaluates to a string
```

Comment:
Is `messageExpression` not type-checked/informationally type-checked the way policy validation rules are? I think there could be situations where trying to eval `messageExpression` fails at runtime. I don't think there's generally too much we could do about it other than emit some kind of "messageExpression evaluation error" message, but it might be worth noting that it could happen.

There could also be potential edge cases where a policy rule validation fails and whatever causes the validation failure also causes the corresponding `messageExpression` to fail to evaluate properly, as opposed to `messageExpression` being invalid no matter what the input is, and that might prove tricky to debug. For example, say we have a rule along the lines of `someProp == "string 1" || someProp == "string 2"` and then have a `messageExpression: "invalid value for someProp:" + someProp`.
If `someProp` does not contain a string for some reason, then it'll result in a runtime error while trying to evaluate `messageExpression` since as far as I know CEL is relatively strict about concatenating. `self + 2 != "some string"` as a CRD validation rule defined on a string property gives me `ERROR: <input>:1:6: found no matching overload for '_+_' applied to '(string, int)'`, for instance.

Those edge cases could be mitigated by wrapping every variable used as part of `matchExpression` in `string()`, i.e, `messageExpression: "invalid value for someProp:" + string(someProp)`, but since that's not always strictly necessary depending on the type of `someProp` it could be a mildly subtle gotcha. We could always have the CEL linter catch that sort of thing, too.

Reply:
Yes, this is all a bit messy when things go wrong.

I was imagining that `messageExpression` would receive treatment almost identical to the `expression` except that any errors (including runtime cost limits) would result in fallback to a auto-generated message, e.g. "expression spec.x != spec.y" and the error from the messageExpression would be logged but not surfaced to the user. I wonder if there is anything better we can do than that?

`messageExpression` should also have a lower runtime cost limit than `expression`. I hadn't thought about that until now.

+1 for linting

Reply:
For auto-generated messages, maybe it would be nice to know what term(s) of the policy rule expression evaluated to false? That could come in handy for debugging. For instance, some kind of error message like:
```
someProp != "badvalue1" && someProp != "badValue2" && someProp != "badValue3"
___________________________^^^^^^^^^^^^^^^^^^^^^^^___________________________
```
It ought to save a few minutes even for simple policy rules, and some headache for more complicated rules where it's not immediately obvious what failed and why. That said, I'm not sure cel-go has the execution tracing that we'd need to implement that kind of thing at the moment, unfortunately.

One simple alternative would be to also print the arguments passed to policy rule that failed. There still might potentially be a little bit of debugging work that users would have to do to figure out what term of the rule failed, but it would be very simple to implement.